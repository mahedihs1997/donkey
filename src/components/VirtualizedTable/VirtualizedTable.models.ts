import { WithStyles } from '@material-ui/core';

import { styles } from './VirtualizedTable.styles';

interface ColumnData {
  dataKey: string;
  label: string;
  numeric?: boolean;
  width: number;
}

export interface Row {
  index: number;
}

export interface MuiVirtualizedTableProps extends WithStyles<typeof styles> {
  columns: ColumnData[];
  headerHeight?: number;
  onRowClick?: () => void;
  rowCount: number;
  rowGetter: (row: Row) => Record<string, string | number>;
  rowHeight?: number;
}
