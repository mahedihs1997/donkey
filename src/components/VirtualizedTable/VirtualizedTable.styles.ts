import { createStyles, Theme } from '@material-ui/core/styles';

export const styles = (theme: Theme) =>
  createStyles({
    flexContainer: {
      display: 'flex',
      alignItems: 'center',
    },
    table: {
      // temporary right-to-left patch, waiting for
      // https://github.com/bvaughn/react-virtualized/issues/454
      '& .ReactVirtualized__Table__headerRow': {
        flip: false,
        paddingRight: theme.direction === 'rtl' ? '0 !important' : undefined,
      },
    },
    tableRow: {
      cursor: 'pointer',
    },
    tableRowHover: {
      '&:hover': {
        backgroundColor: theme.palette.grey[200],
      },
    },
    tableCell: {
      flex: 1,
    },
    noClick: {
      cursor: 'initial',
    },
  });
