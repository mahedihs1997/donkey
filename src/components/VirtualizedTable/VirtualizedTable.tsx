import { withStyles } from '@material-ui/core';
import TableCell from '@material-ui/core/TableCell';
import clsx from 'clsx';
import React from 'react';
import {
  AutoSizer,
  Column,
  Table,
  TableCellRenderer,
  TableHeaderProps,
} from 'react-virtualized';

import { MuiVirtualizedTableProps, Row } from './VirtualizedTable.models';
import { styles } from './VirtualizedTable.styles';

const MuiVirtualizedTable: React.FunctionComponent<MuiVirtualizedTableProps> = (
  props
) => {
  const {
    headerHeight = 48,
    rowHeight = 48,
    classes,
    onRowClick,
    columns,
    ...tableProps
  } = props;

  const getRowClassName = ({ index }: Row) => {
    return clsx(classes.tableRow, classes.flexContainer, {
      [classes.tableRowHover]: index !== -1 && onRowClick != null,
    });
  };

  const cellRenderer: TableCellRenderer = ({ cellData, columnIndex }) => {
    return (
      <TableCell
        component="div"
        className={clsx(classes.tableCell, classes.flexContainer, {
          [classes.noClick]: onRowClick == null,
        })}
        variant="body"
        style={{ height: rowHeight }}
        align={
          (columnIndex != null && columns[columnIndex].numeric) || false
            ? 'right'
            : 'left'
        }
      >
        {cellData}
      </TableCell>
    );
  };

  const headerRenderer = ({
    label,
    columnIndex,
  }: TableHeaderProps & { columnIndex: number }) => {
    return (
      <TableCell
        component="div"
        className={clsx(
          classes.tableCell,
          classes.flexContainer,
          classes.noClick
        )}
        variant="head"
        style={{ height: headerHeight }}
        align={columns[columnIndex].numeric || false ? 'right' : 'left'}
      >
        <span>{label}</span>
      </TableCell>
    );
  };

  return (
    <AutoSizer>
      {({ height, width }) => (
        <Table
          height={height}
          width={width}
          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
          rowHeight={rowHeight!}
          gridStyle={{
            direction: 'inherit',
          }}
          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
          headerHeight={headerHeight!}
          className={classes.table}
          {...tableProps}
          rowClassName={getRowClassName}
        >
          {columns.map(({ dataKey, ...other }, index) => {
            return (
              <Column
                key={dataKey}
                headerRenderer={(headerProps) =>
                  headerRenderer({
                    ...headerProps,
                    columnIndex: index,
                  })
                }
                className={classes.flexContainer}
                cellRenderer={cellRenderer}
                dataKey={dataKey}
                {...other}
              />
            );
          })}
        </Table>
      )}
    </AutoSizer>
  );
};

const VirtualizedTable = withStyles(styles)(MuiVirtualizedTable);

export default VirtualizedTable;
