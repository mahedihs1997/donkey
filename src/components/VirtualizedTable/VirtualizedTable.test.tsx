import React from 'react';
import TestRenderer from 'react-test-renderer';

import VirtualizedTable from '.';

const data = [
  { username: 'Ugendo', name: 'Uga' },
  { username: 'Bugendo', name: 'Buga' },
];

describe('<VirtualizedTable />', () => {
  it('should render without throwing an error and match snapshot', () => {
    const testRenderer = TestRenderer.create(
      <VirtualizedTable
        rowCount={data.length}
        rowGetter={({ index }) => data[index]}
        columns={[
          {
            width: 120,
            label: 'Name',
            dataKey: 'username',
          },
          {
            width: 120,
            label: 'Lies',
            dataKey: 'name',
          },
        ]}
      />
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });
});
