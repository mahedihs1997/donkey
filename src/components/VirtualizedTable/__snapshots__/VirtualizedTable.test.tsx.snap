// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VirtualizedTable /> should render without throwing an error and match snapshot 1`] = `
<div
  style={
    Object {
      "height": 0,
      "overflow": "visible",
      "width": 0,
    }
  }
>
  <div
    aria-colcount={2}
    aria-rowcount={2}
    className="ReactVirtualized__Table MuiVirtualizedTable-table-2"
    role="grid"
    style={Object {}}
  >
    <div
      className="ReactVirtualized__Table__headerRow MuiVirtualizedTable-tableRow-3 MuiVirtualizedTable-flexContainer-1"
      role="row"
      style={
        Object {
          "height": 48,
          "overflow": "hidden",
          "paddingRight": 0,
          "width": 0,
        }
      }
    >
      <div
        className="ReactVirtualized__Table__headerColumn"
        role="columnheader"
        style={
          Object {
            "WebkitFlex": "0 1 120px",
            "flex": "0 1 120px",
            "msFlex": "0 1 120px",
          }
        }
      >
        <div
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiVirtualizedTable-tableCell-5 MuiVirtualizedTable-flexContainer-1 MuiVirtualizedTable-noClick-6 MuiTableCell-alignLeft"
          role="cell"
          style={
            Object {
              "height": 48,
            }
          }
        >
          <span>
            Name
          </span>
        </div>
      </div>
      <div
        className="ReactVirtualized__Table__headerColumn"
        role="columnheader"
        style={
          Object {
            "WebkitFlex": "0 1 120px",
            "flex": "0 1 120px",
            "msFlex": "0 1 120px",
          }
        }
      >
        <div
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiVirtualizedTable-tableCell-5 MuiVirtualizedTable-flexContainer-1 MuiVirtualizedTable-noClick-6 MuiTableCell-alignLeft"
          role="cell"
          style={
            Object {
              "height": 48,
            }
          }
        >
          <span>
            Lies
          </span>
        </div>
      </div>
    </div>
    <div
      aria-label="grid"
      aria-readonly={null}
      className="ReactVirtualized__Grid ReactVirtualized__Table__Grid"
      onScroll={[Function]}
      role="rowgroup"
      style={
        Object {
          "WebkitOverflowScrolling": "touch",
          "boxSizing": "border-box",
          "direction": "inherit",
          "height": -48,
          "overflowX": "hidden",
          "overflowY": "auto",
          "position": "relative",
          "width": 0,
          "willChange": "transform",
        }
      }
      tabIndex={0}
    />
  </div>
</div>
`;
